xaitk_saliency/impls/gen_object_detector_blackbox_sal/occlusion_based.py

```diff
@@ -0,0 +1,172 @@
+import numpy as np
+from typing import Optional, Union, Sequence, Dict, Hashable, Iterable, Tuple, Any, Type, TypeVar
+
+from xaitk_saliency.interfaces.gen_object_detector_blackbox_sal import GenerateObjectDetectorBlackboxSaliency
+from xaitk_saliency.interfaces.perturb_image import PerturbImage
+from xaitk_saliency.interfaces.gen_detector_prop_sal import GenerateDetectorProposalSaliency
+from xaitk_saliency.utils.detection import format_detection
+from xaitk_saliency.utils.masking import occlude_image_streaming
+
+from smqtk_detection.interfaces.detect_image_objects import DetectImageObjects
+from smqtk_detection.utils.bbox import AxisAlignedBoundingBox
+from smqtk_core.configuration import (
+    from_config_dict,
+    make_default_config,
+    to_config_dict,
+)
+
+C = TypeVar("C", bound="PerturbationOcclusion")
+
+
+class PerturbationOcclusion (GenerateObjectDetectorBlackboxSaliency):
+    """
+    Generator composed of modular perturbation and occlusion-based algorithms.
+
+    This implementation exposes a public attribute `fill`.
+    This may be set to a scalar or sequence value to indicate a color that
+    should be used for filling occluded areas as determined by the given
+    `PerturbImage` implementation.
+    This is a parameter to be set during runtime as this is most often driven
+    by the black-box algorithm used, if at all.
+
+    :param perturber: `PerturbImage` implementation instance for generating
+        occlusion masks.
+    :param generator: `GenerateDetectorProposalSaliency` implementation
+        instance for generating saliency masks given occlusion masks and
+        black-box detector outputs.
+    :param threads: Optional number threads to use to enable parallelism in
+        applying perturbation masks to an input image. If 0, a negative value,
+        or `None`, work will be performed on the main-thread in-line.
+    """
+
+    def __init__(
+        self,
+        perturber: PerturbImage,
+        generator: GenerateDetectorProposalSaliency,
+        threads: Optional[int] = 0
+    ):
+        self._perturber = perturber
+        self._generator = generator
+        self._threads = threads
+        # Optional fill color
+        self.fill: Optional[Union[int, Sequence[int]]] = None
+
+    def _generate(
+        self,
+        ref_image: np.ndarray,
+        bboxes: np.ndarray,
+        scores: np.ndarray,
+        blackbox: DetectImageObjects,
+        objectness: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        ref_dets_mat = format_detection(bboxes, scores, objectness)
+
+        pert_masks = self._perturber(ref_image)
+
+        pert_dets = blackbox.detect_objects(
+            occlude_image_streaming(
```

Comment:
Did we want to use the new threaded "batch" version here now instead? Or does that still use too much RAM with the quantity of perturbations used?

Reply:
I did some short tests of both and found that the difference in RAM usage was almost nothing, even with a high number of masks, so I switched to the batched version.